retrieval/targetmanager_test.go

```diff
@@ -12,3 +12,63 @@
 // limitations under the License.
 
 package retrieval
+
+import (
+	"testing"
+
+	"golang.org/x/net/context"
+	"gopkg.in/yaml.v2"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/storage/local"
+)
+
+func TestTargetSetRecreatesTargetGroupsEveryRun(t *testing.T) {
+	scrapeConfig := &config.ScrapeConfig{}
+
+	sOne := `
+job_name: "foo"
+dns_sd_configs:
+- names:
+  - "srv.name.one.example.org"
+`
+	if err := yaml.Unmarshal([]byte(sOne), scrapeConfig); err != nil {
+		t.Fatalf("Unable to load YAML config sOne: %s", err)
+	}
+
+	// Not properly setting it up, but that seems okay
+	mss := &local.MemorySeriesStorage{}
+
+	ts := newTargetSet(scrapeConfig, mss)
+
+	ts.runProviders(context.Background(), providersFromConfig(scrapeConfig))
+
+	verifyPresence(t, ts.tgroups, "dns/0/srv.name.one.example.org", true)
+
+	sTwo := `
+job_name: "foo"
+dns_sd_configs:
+- names:
+  - "srv.name.two.example.org"
+`
+	if err := yaml.Unmarshal([]byte(sTwo), scrapeConfig); err != nil {
+		t.Fatalf("Unable to load YAML config sTwo: %s", err)
+	}
+
+	ts.runProviders(context.Background(), providersFromConfig(scrapeConfig))
+
+	verifyPresence(t, ts.tgroups, "dns/0/srv.name.one.example.org", false)
+	verifyPresence(t, ts.tgroups, "dns/0/srv.name.two.example.org", true)
+
+}
+
+func verifyPresence(t *testing.T, tgroups map[string][]*Target, name string, present bool) {
```

Comment:
For such a specific function that's not shared across tests, it might be more convenient to inline it into the actual test function, so we don't run into name collisions etc.